compiler/front_end/parser.py

```diff
@@ -12,120 +12,42 @@
-            tokens, errors = tokenizer.tokenize(example, "")
-            if errors:
-                raise ParserGenerationError(str(errors))
-
-            for i in range(len(tokens)):
-                if tokens[i].symbol == "BadWord" and tokens[i].text == "$ANY":
-                    tokens[i] = lr1.ANY_TOKEN
-
-            error_token = None
-            for i in range(len(tokens)):
-                if tokens[i].symbol == "BadWord" and tokens[i].text == "$ERR":
-                    error_token = tokens[i + 1]
-                    del tokens[i]
-                    break
-            else:
-                raise ParserGenerationError(
-                    "No error token marker '$ERR' in:\n" + error_example
-                )
-
-            result.append((tokens, error_token, message.strip(), example))
-    return result
-
-
-def generate_parser(start_symbol, productions, error_examples):
-    """Generates a parser from grammar, and applies error_examples.
-
-    Arguments:
-        start_symbol: the start symbol of the grammar (a string)
-        productions: a list of parser_types.Production in the grammar
-        error_examples: A list of (source tokens, error message, source text)
-            tuples.
-
-    Returns:
-        A parser.
-
-    Raises:
-        ParserGenerationError: There is a problem generating the parser.
-    """
-    parser = lr1.Grammar(start_symbol, productions).parser()
-    if parser.conflicts:
-        raise ParserGenerationError("\n".join([str(c) for c in parser.conflicts]))
-    for example in error_examples:
-        mark_result = parser.mark_error(example[0], example[1], example[2])
-        if mark_result:
-            raise ParserGenerationError(
-                "error marking example: {}\nExample:\n{}".format(
-                    mark_result, example[3]
-                )
-            )
-    return parser
-
-
 @simple_memoizer.memoize
-def _load_module_parser():
-    error_examples = parse_error_examples(
-        resources.load("compiler.front_end", "error_examples")
-    )
-    return generate_parser(
-        module_ir.START_SYMBOL, module_ir.PRODUCTIONS, error_examples
-    )
-
-
-@simple_memoizer.memoize
-def _load_expression_parser():
-    return generate_parser(module_ir.EXPRESSION_START_SYMBOL, module_ir.PRODUCTIONS, [])
+def _load_parsers():
+    module_parser = cached_parser.module_parser()
+    expression_parser = cached_parser.expression_parser()
+    if (
+        module_parser.productions
+        == set(module_ir.PRODUCTIONS)
+        | {parser_types.Production(lr1.START_PRIME, (module_ir.START_SYMBOL,))}
+    ) and (
+        expression_parser.productions
+        == set(module_ir.PRODUCTIONS)
+        | {
+            parser_types.Production(
+                lr1.START_PRIME, (module_ir.EXPRESSION_START_SYMBOL,)
+            )
+        }
+    ):
+        return module_parser, expression_parser
+    return make_parser.build_module_parser(), make_parser.build_expression_parser()
```

Comment:
Could be worthwhile to log when we aren't using the cached parsers so that

1. The contributor has a confirmation that they're using the parser with their new changes and not the cached one and
2. It can be a reminder to regenerate the pre-generated parser before committing.

Reply:
I will look into this, but I need to actually set up compiler warnings first, which I would prefer to do in a separate PR.

Reply:
That makes sense, SGTM

Reply:
+1 as a follow up